R/read-gs.R

```diff
@@ -18,14 +19,225 @@
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    
+    
+    
+    return(read_gradescope_grades(grades))
+      
+  } else if (source == "Canvas"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    # the read_canvas_grades does all of the lifting
+    return(read_canvas_grades(grades))
+    
+  } else {
+    stop(paste0(
+      "Specified source type ", source,
+      " is not currently supported."
+    ))
+  }
+  
+  
+ 
+}
+
+read_gradescope_grades <- function(df){
+  # almost identity function when reading in gradescope grades 
+  # to create symmetry with read_canvas_grades
+  
+  attr(df, "source") <- "Gradescope"
+  check_data_format(df)
+}
+
+
+
+#' @importFrom stringr str_extract str_match str_replace
+#' @importFrom dplyr filter select slice rename_with rename bind_cols mutate_at
+#' @importFrom purrr discard
+#' @importFrom lubridate make_difftime
+#' 
+read_canvas_grades <- function(grades){
+  # Convert Canvas grade dataframe into standard format
+  #
+  # Takes in a dataframe uploaded from Canvas. 
+  # Returns a dataframe in our standardized format (ie like Gradescope).
+  #
+  #
+  
+  # first determine which columns are assignments
+  
+  # canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+  # coerce into numerical and convert excused assignments into set value
+  grades <- grades |>
+    dplyr::filter(!is.na(Student)) |>
+    dplyr::mutate_at(assignments, function(x) {
+        dplyr::if_else(x == "EX", NA, x) |>
+        as.numeric()
+      })
+  
+ 
+  # now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student %in% c( "    Points Possible", "Points Possible")) |>
+    dplyr::select(all_of(assignments)) |>
+    dplyr::slice(rep.int(1, nrow(grades))) |>
+    dplyr::rename_with(
+        function(x){
+          paste0(x, " - Max Points")
+        }
+    )
+  
+  
+  # remove unneeded cols, add max_point column, and remove rows that are not students
+  grades <- grades |>
+    dplyr::select(c(all_of(assignments), "ID", "Student", 
+                    "SIS User ID", "Section")) |>
+    dplyr::bind_cols(max_points) |>
+    dplyr::filter(!(Student %in% c("    Points Possible", "Student, Test", "Points Possible", NA)))
+  
+  # add columns for submission time and lateness
+  
+  sub_cols <- paste0(assignments, " - Submission Time")
+  
+  late_cols <- paste0(assignments, " - Lateness (H:M:S)")
+  
+  grades[sub_cols] <- NA
+  
+  grades[late_cols] <- lubridate::make_difftime(NA)
```

Comment:
Shouldn't default be "00:00:00"? I don't remember if we have drop_na in lateness